GowerMetric.py

```diff
@@ -0,0 +1,361 @@
+        elif self.ratio_scale_normalization == 'iqr' or self.ratio_scale_normalization == 'kde':
+            q1, q3 = np.percentile(ratio_cols, [25, 75], axis=0)
+            self.ranges_[self.ratio_scale_idx] = q3 - q1
+
+            zero_ranges_idx = (self.ranges_ == 0) & (self.ratio_scale_idx)
+            self.ranges_[zero_ranges_idx] = np.ptp(X[:, zero_ranges_idx], axis=0)
+
+        if self.ratio_scale_normalization == 'kde':
+            n = X.shape[0]
+            c = 1.06
+
+            s = np.std(ratio_cols, axis=0)
+            self.h_[self.ratio_scale_idx] = (
+                c / n ** (1 / 5) * np.min([s, self.ranges_[self.ratio_scale_idx] / 1.34], axis=0)
+            )
+
+        loader = GowerMetricWeights(self)
+        if self.weights == "precomputed":
+            loader.load_weights(self.precomputed_weights_file)
+        elif self.weights == "cpcc":
+            self.weights = np.ones(self.n_features_in_)
+            loader.select_weights(X)
+
+        loader.select_number_of_clusters(X)
+
+    def __call__(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+    ):
+        # TODO - no check for fit
+        # sklearn.utils.validation.check_is_fitted(self)
+        assert self.n_features_in_ == len(vector_1) | len(vector_2)
+
+        # Distance for cat_nom and bin_sym columns
+        if self.cat_nom_num > 0:
+            cat_nom_cols_1 = vector_1[self.cat_nom_idx]
+            cat_nom_cols_2 = vector_2[self.cat_nom_idx]
+            cat_nom_dist = 1.0 - (cat_nom_cols_1 == cat_nom_cols_2)
+
+            # Weighted variant - dot product with according self.weights columns
+            # Unweighted - a sum of distances between all columns
+            if self.weights is not None:
+                cat_nom_dist = cat_nom_dist @ self.weights[self.cat_nom_idx]
+            else:
+                cat_nom_dist = cat_nom_dist.sum()
+        else:
+            cat_nom_dist = 0.0
+
+        # Distance for bin_asym columns
+        if self.bin_asym_num > 0:
+            bin_asym_cols_1 = vector_1[self.bin_asym_idx]
+            bin_asym_cols_2 = vector_2[self.bin_asym_idx]
+
+            zeros = (bin_asym_cols_1 == 0) | (bin_asym_cols_2 == 0)
+
+            bin_asym_dist = (bin_asym_cols_1 != bin_asym_cols_2) | zeros
+
+            if self.weights is not None:
+                bin_asym_dist = bin_asym_dist @ self.weights[self.bin_asym_idx]
+            else:
+                bin_asym_dist = bin_asym_dist.sum()
+        else:
+            bin_asym_dist = 0.0
+
+        # Distance for ratio_scale columns
+        if self.ratio_scale_num > 0:
+            ratio_scale_cols_1 = vector_1[self.ratio_scale_idx]
+            ratio_scale_cols_2 = vector_2[self.ratio_scale_idx]
+            ratio_dist = np.abs(ratio_scale_cols_1 - ratio_scale_cols_2)
+
+            above_threshold = ratio_dist >= self.ranges_[self.ratio_scale_idx]
+            below_threshold = ratio_dist <= self.h_[self.ratio_scale_idx]
```

Comment:
Remember that `self.h_` is used only for KDE normalization, this variable shouldn't even exist otherwise